### ISSUE TITLE
TinyTex status 127 error 

### ISSUE THREAD

**Issue by karstendevries** [yihui/tinytex#79]
Hi Yihui,

We are trying to install and run tinytex on our Windows 10 notebooks, but we ran into the following errors:

---
"C:/Program Files/RStudio/bin/pandoc/pandoc" +RTS -K512m -RTS karsten.utf8.md --to latex --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output karsten.tex --template "C:\PROGRA~1\R\R-34~1.1\library\RMARKD~1\rmd\latex\DEFAUL~3.TEX" --highlight-style tango --latex-engine pdflatex --variable graphics=yes --variable "geometry:margin=1in" --variable "compact-title:yes" 
output file: karsten.knit.md

Error: Failed to compile karsten.tex.
In addition: Warning messages:
1: running command '"pdflatex" -halt-on-error -interaction=batchmode "karsten.tex"' had status 127 
2: running command '"pdflatex" -halt-on-error -interaction=batchmode "karsten.tex"' had status 127 
Execution halted

No TeX installation detected (TeX is required to create PDF output). You should install a recommended TeX distribution for your platform:

  Windows: MiKTeX (Complete) - http://miktex.org/2.9/setup
  (NOTE: Be sure to download the Complete rather than Basic installation)

  Mac OS X: TexLive 2013 (Full) - http://tug.org/mactex/
  (NOTE: Download with Safari rather than Chrome _strongly_ recommended)

  Linux: Use system package manager
---

- tinytex:::is_tinytex() results in:
TRUE

- xfun::session_info('tinytex') results in:
R version 3.4.1 (2017-06-30)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200), RStudio 1.0.143
Locale:
  LC_COLLATE=Dutch_Netherlands.1252  LC_CTYPE=Dutch_Netherlands.1252    LC_MONETARY=Dutch_Netherlands.1252 LC_NUMERIC=C                      
  LC_TIME=Dutch_Netherlands.1252    
Package version:
  tinytex_0.8 tools_3.4.1 xfun_0.3 

- tinytex::tinytex_root() results in: 
[1] "C:\\Users\\vriesk1\\AppData\\Roaming\\TinyTeX"

- we added the following options:
options(tinytex.tlmgr.path = 'C:\\Users\\vriesk1\\AppData\\Roaming\\TinyTeX\\bin\\win32\\tlmgr.bat')
Sys.setenv(PATH = paste(Sys.getenv("PATH"), "C:\\Users\\vriesk1\\AppData\\Roaming\\TinyTeX", sep=.Platform$path.sep))

- our (test) Rmd document looks like:
---
title: "TESTJE"
author: "Noani, Margreet en Karsten"
date: "15 oktober 2018"
output:
  pdf_document: default
  word_document: default
---

```{r setup, include=FALSE}
options(tinytex.verbose = TRUE)
```
## R Markdown

In addition to the above we want to mention that we don't use MixTex or any other Latex packages :)
Do you have any idea what the 127 status mean and how we can fix it?

Thanks in advance for answering our question!

Kind regards,
Karsten en Noani

**Comment by yihui** [comment]
Did you restart RStudio after installing TinyTeX? What do `Sys.which('pdflatex')` and `system2('pdflatex', '--version')` return? In theory, you shouldn't need to change the `tinytex.tlmgr.path` option or tweak the `PATH` variable on Windows.

**Comment by karstendevries** [comment]
Hi Yihui,

Thank you for answering so quickly! 

Sys.which('pdflatex') resulted in "C:\\Users\\vriesk1\\AppData\\Roaming\\TinyTeX\\bin\\win32\\pdflatex.exe", while the system2 command resulted in "Warning message: running command '"pdflatex" --version' had status 127".

**Comment by yihui** [comment]
I wonder if `cmd.exe` can be found from PATH: `Sys.which('cmd')`. Could you provide your `Sys.getenv()`? You can remove any output that may contain sensitive information.

**Comment by karstendevries** [comment]
Yes, of course! Sys.which('cmd') returns in "C:\\windows\\SYSTEM32\\cmd.exe" and Sys.getenv() results in:

ALLUSERSPROFILE                                  C:\ProgramData
APPDATA                                          C:\Users\vriesk1\AppData\Roaming
CommonProgramFiles                               C:\Program Files\Common Files
CommonProgramFiles(x86)                          C:\Program Files (x86)\Common Files
CommonProgramW6432                               C:\Program Files\Common Files
COMPUTERNAME                                     << removed :) >>
ComSpec                                          C:\windows\system32\cmd.exe
DEFLOGDIR                                        << removed :) >>
DISPLAY                                          :0
GFORTRAN_STDERR_UNIT                             -1
GFORTRAN_STDOUT_UNIT                             -1
HOME                                             << removed :) >>
HOMEDRIVE                                        C:
HOMEPATH                                         \Users\vriesk1
LOCALAPPDATA                                     C:\Users\vriesk1\AppData\Local
LOGONSERVER                                      << removed :) >>
NUMBER_OF_PROCESSORS                             4
OneDrive                                         << removed :) >>
OneDriveSync                                     << removed :) >>
OS                                               Windows_NT
PATH                                             C:\Program Files\R\R-3.4.1\bin\x64;C:\Program Files\RStudio\bin\;C:\Program Files (x86)\Common
                                                 Files\Oracle\Java\javapath;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\Program
                                                 Files\Intel\WiFi\bin\;C:\Program Files\Common
                                                 Files\Intel\WirelessCommon\;C:\Users\vriesk1\AppData\Local\Microsoft\WindowsApps;
PATHEXT                                          .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PROCESSOR_ARCHITECTURE                           AMD64
PROCESSOR_IDENTIFIER                             Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
PROCESSOR_LEVEL                                  6
PROCESSOR_REVISION                               8e09
ProgramData                                      C:\ProgramData
ProgramFiles                                     C:\Program Files
ProgramFiles(x86)                                C:\Program Files (x86)
ProgramW6432                                     C:\Program Files
PSModulePath                                     C:\Program
                                                 Files\WindowsPowerShell\Modules;C:\windows\system32\WindowsPowerShell\v1.0\Modules;C:\Program
                                                 Files (x86)\Microsoft Azure Information Protection\Powershell
PUBLIC                                           C:\Users\Public
R_ARCH                                           /x64
R_COMPILED_BY                                    gcc 4.9.3
R_DOC_DIR                                        C:/PROGRA~1/R/R-34~1.1/doc
R_HOME                                           C:/PROGRA~1/R/R-34~1.1
R_LIBS_USER                                      C:\Program Files\R\R-3.4.1\library
R_USER                                           << removed :) >>
RMARKDOWN_MATHJAX_PATH                           C:/Program Files/RStudio/resources/mathjax-26
RS_LOCAL_PEER                                    \\.\pipe\10544-rsession
RS_RPOSTBACK_PATH                                C:/Program Files/RStudio/bin/rpostback
RS_SHARED_SECRET                                 63341846741
RSTUDIO                                          1
RSTUDIO_MSYS_SSH                                 C:/Program Files/RStudio/bin/msys-ssh-1000-18
RSTUDIO_PANDOC                                   C:/Program Files/RStudio/bin/pandoc
RSTUDIO_SESSION_PORT                             10544
RSTUDIO_USER_IDENTITY                            vriesk1
RSTUDIO_WINUTILS                                 C:/Program Files/RStudio/bin/winutils
SESSIONNAME                                      Console
SystemDrive                                      C:
SystemRoot                                       C:\windows
TEMP                                             C:\Users\vriesk1\AppData\Local\Temp
TMP                                              C:\Users\vriesk1\AppData\Local\Temp
UATDATA                                          C:\windows\CCM\UATData\D9F8C395-CAB8-491d-B8AC-179A1FE1BE77
USERDNSDOMAIN                                    << removed :) >>
USERDOMAIN                                       << removed :) >>
USERDOMAIN_ROAMINGPROFILE                        << removed :) >>
USERNAME                                         vriesk1
USERPROFILE                                      C:\Users\vriesk1
windir                                           C:\windows

**Comment by yihui** [comment]
I wonder why `C:\Users\vriesk1\AppData\Roaming\TinyTeX\bin\win32` was not present in `PATH`, but at the same time you said

> `Sys.which('pdflatex')` resulted in `"C:\Users\vriesk1\AppData\Roaming\TinyTeX\bin\win32\pdflatex.exe"`

The two facts contradict each other.

The way to add TinyTeX to PATH is `tinytex::tlmgr_path()` (need to restart RStudio if the command can be run successfully).

**Comment by karstendevries** [comment]
Thank you! We've updated the R packages with the command mentioned above. With regard to the tinytex::tlmgr_path command; we don't have admin privileges to run it, is there another way to add TinyTeX to PATH? :)

**Comment by yihui** [comment]
TinyTeX doesn't require admin privileges. On Windows, there are two types of PATH variables: the one for your system, and the one only for yourself. You can modify the latter without admin privileges.

I assume you have run `tinytex::tlmgr_path()` and it couldn't modify the PATH variable?

**Comment by karstendevries** [comment]
Yes, we tried to run tinytex::tlmgr_path(). Is there a specific command to modify the user PATH variable? :) In addition, does TinyTeX also not require admin privileges in order to run pdflatex.exe when generating PDF documents?

**Comment by karstendevries** [comment]
tinytex::tlmgr_path() retuns:

tlmgr path add
Dit programma wordt geblokkeerd door Groepsbeleid. Neem voor meer informatie contact op met de systeembeheerder.
Warning message:
running command '"tlmgr" path add' had status 1 

Based on the above we think that our IT department has blocked the system command setx. We also couldn't modify the (system) PATH variable using the link you provided. However, we can add it to our accounts specifically causing the "No TeX installation detected" error to disappear.

Unfortunately, we still get the following errors when trying to Knit a PDF document:

output file: asdf.knit.md

Error: Failed to compile asdf.tex.
In addition: Warning messages:
1: running command '"pdflatex" -halt-on-error -interaction=batchmode "asdf.tex"' had status 127 
2: running command '"pdflatex" -halt-on-error -interaction=batchmode "asdf.tex"' had status 127 
Execution halted

Given the fact that our IT department blocked the "tlmgr path add" command, could it be that they also block pdflatex.exe? If so, is there still a way to generate PDF's using TinyTex? :)

**Comment by yihui** [comment]
Oh, IT... Finally [the worst part](https://yihui.name/en/2018/01/limit-of-bureaucracy/) in problem-solving has come :) What is the point of blocking the command `setx`...

> we can add it to our accounts specifically causing the "No TeX installation detected" error to disappear

Do you mean you can add the path of TinyTeX to the *User* PATH Variable? Users without admin privileges can't modify the system PATH variable. That's normal.

What does the following code return?

```r
Sys.which('pdflatex')
system('pdflatex --version')
shell('pdflatex --version')
```

**Comment by karstendevries** [comment]
> Oh, IT... Finally the worst part in problem-solving has come :) What is the point of blocking the command setx...

Haha, I couldn't agree more! 

Yes, we can add the path of TinyTex to the User PATH variable. 

Sys.which('pdflatex'):
"C:\\Users\\vriesk1\\AppData\\Roaming\\TinyTeX\\bin\\win32\\pdflatex.exe" 

system('pdflatex --version'):
Warning message:
running command 'pdflatex --version' had status 127 

(generating or knitting PDF documents also gives a 127 status)

shell('pdflatex --version'):
Dit programma wordt geblokkeerd door Groepsbeleid. Neem voor meer informatie contact op met de systeembeheerder.
Warning messages:
1: running command 'C:\windows\system32\cmd.exe /c pdflatex --version' had status 1 
2: In shell("pdflatex --version") :
  'pdflatex --version' execution failed with error code 1

Thank you so much for all your help! :)

**Comment by yihui** [comment]
How about `processx::run('pdflatex', '--version')`?

**Comment by karstendevries** [comment]
processx::run('pdflatex', '--version') returns:

Error in process_initialize(self, private, command, args, stdin, stdout,  : 
  processx error, create process: #1260 Dit programma wordt geblokkeerd door Groepsbeleid. Neem voor meer informatie contact op met de systeembeheerder.
 at 'win/processx.c:1039'

I think this one is blocked as well.

**Comment by yihui** [comment]
I see. Then there isn't anything I could do on my side, since your IT seems to lack the sense of existence (like all other IT in the world), which I can't satisfy... 😞

**Comment by karstendevries** [comment]
We understand... no problem, thank you so much for all your help! Hopefully we will find a workaround as we still need to generate PDF documents :)

**Comment by karstendevries** [comment]
Wow, that sounds promising! Given that there already is a github account for this new project, is it already available and ready for use? Luckily our IT supports Chrome, so this could be a good workaround!! :)

**Comment by yihui** [comment]
It depends on what kind of features you want in PDF. As I said, the project is still at a very early stage, and there may be many rough edges. The other problem is that there is not enough documentation yet. You may take a look at https://rstudio.github.io/pagedown, and decide whether it looks good enough to you.
